cmd/migrate/main.go

```diff
@@ -0,0 +1,161 @@
+
+	// Get downstreamMachines
+	downstreamMachines, err := downstreamClient.MachineV1alpha1().Machines().List(metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list downstream machines: %v", err)
+	}
+
+	// Convert the machine, create the new machine, delete the old one, wait for it to be absent
+	// We do this in one loop to avoid ending up having all machines in  both the new and the old format if deletion
+	// failes for whatever reason
+	for _, downstreamMachine := range downstreamMachines.Items {
+		convertedClusterv1alpha1Machine, err := conversions.ConvertV1alpha1DownStreamMachineToV1alpha1ClusterMachine(downstreamMachine)
+		if err != nil {
+			return fmt.Errorf("failed to convert machine %s: %v", downstreamMachine.Name, err)
+		}
+
+		createdClusterV1alpha1Machine, err := clusterv1alpha1Client.ClusterV1alpha1().Machines(convertedClusterv1alpha1Machine.Namespace).Create(convertedClusterv1alpha1Machine)
+		if err != nil {
+			return fmt.Errorf("failed to create clusterv1alpha1.machine %s: %v", convertedClusterv1alpha1Machine.Name, err)
+		}
+
+		node, err := kubeClient.CoreV1().Nodes().Get(convertedClusterv1alpha1Machine.Spec.Name, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("Failed to get node %s for machine %s: %v", convertedClusterv1alpha1Machine.Spec.Name, convertedClusterv1alpha1Machine.Name, err)
+		}
+		for idx, ownerRef := range node.OwnerReferences {
+			if ownerRef.UID == downstreamMachine.UID {
+				node.OwnerReferences = append(node.OwnerReferences[:idx], node.OwnerReferences[idx+1:]...)
+				break
+			}
+		}
+		newOwnerRef := node.OwnerReferences
+		gv := clusterv1alpha1.SchemeGroupVersion
+		newOwnerRef = append(newOwnerRef, *metav1.NewControllerRef(createdClusterV1alpha1Machine, gv.WithKind("Machine")))
+
+		// We retry this because nodes get frequently updated so there is a reasonable chance this may fail
+		if err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			node, err := kubeClient.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			node.OwnerReferences = newOwnerRef
+			if _, err := kubeClient.CoreV1().Nodes().Update(node); err != nil {
+				return err
+			}
+			return nil
+		}); err != nil {
+			return fmt.Errorf("failed to update OwnerRef on node %s: %v", node.Name, err)
+		}
+
+		finalizers := sets.NewString(downstreamMachine.Finalizers...)
+		finalizers.Delete(controller.FinalizerDeleteInstance)
+		downstreamMachine.Finalizers = finalizers.List()
+		if _, err := downstreamClient.MachineV1alpha1().Machines().Update(&downstreamMachine); err != nil {
+			return fmt.Errorf("failed to update downstream machine %s after removing finalizer: %v", convertedClusterv1alpha1Machine.Name, err)
+		}
+		if err := downstreamClient.MachineV1alpha1().Machines().Delete(convertedClusterv1alpha1Machine.Name, nil); err != nil {
+			return fmt.Errorf("failed to delete machine %s: %v", convertedClusterv1alpha1Machine.Name, err)
```

Comment:
deleting the machine resource will effectively delete the provider instance.

Reply:
Only if a machine-controller is running. I will wrap the actual migration with  a leader-election to ensure no controller runs and no other migrator runs.

Reply:
You're just deleting one finalizer, but you actually need to migrate & delete all finalizers.

Otherwise you cannot guarantee that the machine gets deleted.
Also you should add a check afterwards to ensure that the machine resource is gone

Reply:
Nevermind check exists

Reply:
Other thing, maybe you should set the `controllerNameAnnotationKey` to ensure that no controller will process it.
You cannot guarantee that no controller runs concurrently -> `RollingUpdate` scenario